transport/x/grpc/response_writer.go

```diff
@@ -24,28 +24,57 @@ import (
 	"bytes"
 
 	"go.uber.org/yarpc/api/transport"
+	"go.uber.org/yarpc/internal/bufferpool"
 	"google.golang.org/grpc/metadata"
 )
 
 type responseWriter struct {
-	*bytes.Buffer
+	buffer             *bytes.Buffer
 	md                 metadata.MD
 	isApplicationError bool
 }
 
-func newResponseWriter(md metadata.MD) *responseWriter {
-	return &responseWriter{
-		bytes.NewBuffer(nil),
-		md,
-		false,
+func newResponseWriter() *responseWriter {
+	return &responseWriter{}
+}
+
+func (r *responseWriter) Write(p []byte) (int, error) {
+	if r.buffer == nil {
+		r.buffer = bufferpool.Get()
 	}
+	return r.buffer.Write(p)
 }
 
 func (r *responseWriter) AddHeaders(headers transport.Headers) {
+	if r.md == nil {
+		r.md = metadata.New(nil)
+	}
 	// TODO: handle error
 	_ = addApplicationHeaders(r.md, headers)
 }
 
 func (r *responseWriter) SetApplicationError() {
 	r.isApplicationError = true
 }
+
+func (r *responseWriter) AddSystemHeader(key string, value string) {
+	if r.md == nil {
+		r.md = metadata.New(nil)
+	}
+	// TODO: handle error
+	_ = addToMetadata(r.md, key, value)
```

Comment:
Not worth handling in this PR?

Reply:
This one is a weird one, not worth handling here - this will only error if you specify a duplicate, which we might just want to allow https://github.com/yarpc/yarpc-go/blob/dev/transport/x/grpc/headers.go#L192